### ISSUE TITLE
Feature Request: Battery Optimization

### ISSUE THREAD

**Issue by mikeaumiller** [JurajNyiri/HomeAssistant-Tapo-Control#457]
**Is your feature request related to a problem? Please describe.**
When the integration is configured for a battery-powered outdoor camera (TC85) it drains the battery within 12 hours. When the integration is disabled the battery remains at a constant state. It appears there's some type of constant connection of some sort to the device.

**Describe the solution you'd like**
Would like the ability to limit how often any connection is made to a Tapo battery-powered camera to pull metrics or make a RTSP connection unless the camera card is clicked start a live stream.

**Describe alternatives you've considered**

**Additional context**
I have (3) Tapo TC85 outdoor cameras and noticed the battery draining 2% per hour when the Tapo-Control integration is enabled . For my dashboard, I have configured my custom card with WebRTC Camera which leverages [go2rtc] and have the initial stream set to "Paused by default" which only display a screenshot until you enable a live stream, but nothing has stopped the battery levels from dropping quickly.

**Comment by JurajNyiri** [comment]
Does this camera use a hub or is it directly connected to wifi?

**Comment by mikeaumiller** [comment]
These camera’s are wifi connected

**Comment by JurajNyiri** [comment]
Do you have onvif enabled (motion detection)? If so, try disabling it and see if it improves or not.

**Comment by mikeaumiller** [comment]
yeah, I thought about that as well... I'll disable (motion detection) and won't configure any camera card and will monitor battery usage over the next couple of hours.

**Comment by JurajNyiri** [comment]
In any case, integration does load up to date data every 30s in one request, we can tone it down, ideally automatically. Do you have battery sensor exposed in HA? Could you copy paste all the camera entity properties here?

**Comment by mikeaumiller** [comment]
I like the idea of automatically or if possible even give the user control of the heartbeat value with a slider or pulldown with predefined values

Below is a screenshot of the battery metric usage. I can run it longer if needed.

![Tapo Camera TC85 Battery Metric](https://github.com/JurajNyiri/HomeAssistant-Tapo-Control/assets/43655754/82dc0bf3-1620-4883-a160-c50879ab46bc)

**Comment by JurajNyiri** [comment]
> Could you copy paste all the camera entity properties here?

**Comment by mikeaumiller** [comment]
Here you go...
![Tapo TC85 - Device Info](https://github.com/JurajNyiri/HomeAssistant-Tapo-Control/assets/43655754/5b39a562-5c0d-4821-8493-44f8135aa421)

![Tapo TC85 - Configuration](https://github.com/JurajNyiri/HomeAssistant-Tapo-Control/assets/43655754/8dad0cc6-b7e4-4a0d-a980-ef8cc64f76a0)

![Tapo TC85 - Diagnostics](https://github.com/JurajNyiri/HomeAssistant-Tapo-Control/assets/43655754/0816e276-d9fe-4e3f-808e-39e856e46a23)

**Comment by JurajNyiri** [comment]
Thank you, could you please provide Attributes of the camera entity? You can find these in `/developer-tools/state`

**Comment by mikeaumiller** [comment]
Here you go...

<img width="1767" alt="image" src="https://github.com/JurajNyiri/HomeAssistant-Tapo-Control/assets/43655754/df78c1c9-ff4c-44b4-b62b-3f7c6ec27799">

<img width="1752" alt="image" src="https://github.com/JurajNyiri/HomeAssistant-Tapo-Control/assets/43655754/0152e1a0-4308-4973-8c84-6cf6bece0ea7">

<img width="1764" alt="image" src="https://github.com/JurajNyiri/HomeAssistant-Tapo-Control/assets/43655754/6a984985-cca2-4398-b750-59b8df3d6191">

<img width="1759" alt="image" src="https://github.com/JurajNyiri/HomeAssistant-Tapo-Control/assets/43655754/df774a99-6906-4724-9b54-9f90e0e8ada0">

<img width="1762" alt="image" src="https://github.com/JurajNyiri/HomeAssistant-Tapo-Control/assets/43655754/6115cf0b-9131-4cee-9328-60568af486a7">

<img width="1752" alt="image" src="https://github.com/JurajNyiri/HomeAssistant-Tapo-Control/assets/43655754/a9ea31c6-dd3e-487d-9b8b-d20c9bd4bcb5">

<img width="1753" alt="image" src="https://github.com/JurajNyiri/HomeAssistant-Tapo-Control/assets/43655754/78001c8f-298d-4b86-9ecc-b23d0c1cc676">

<img width="1761" alt="image" src="https://github.com/JurajNyiri/HomeAssistant-Tapo-Control/assets/43655754/c621fedb-afd1-4f46-a7e4-1ee6d2714d96">

**Comment by mikeaumiller** [comment]
Maybe this attribute under the HD stream:            model_name: TC85

**Comment by JurajNyiri** [comment]
Thank you I now have everything I need to add support.
We can use `power` value `BATTERY` along with detecting whether it is a child device or not via pytapo to determine it is a wifi camera on battery without a hub.

**Comment by mikeaumiller** [comment]
Ahh.. yeah that's a better idea.

**Comment by ion-storm** [comment]
Hey hey! I have these cameras also the TC85 model, will be willing to send the dev out a camera to expand support, reach me on twitter @ionstorm

**Comment by ion-storm** [comment]
The TC85 doesn't appear to have ONVIF, it has control port 443, rtsp 554, and a http video port on 8800, unsure of all the capabilities, there could be other ways to get the motion events.  I also am noticing that the battery drain is 2% per hour at least with the plugin enabled.  All other functionality works except motion detection.

**Comment by mikeaumiller** [comment]
Just circling back to see if there's any update on this original request to optimize battery-powered cameras?

**Comment by Exelord** [comment]
Same issue noticed for C425 which is TC85. 10% battery drain per day with integration turned on. Only webhooks enabled

**Comment by jcconnell** [comment]
> Same issue noticed for C425 which is TC85. 10% battery drain per day with integration turned on. Only webhooks enabled

Somewhat unrelated - I'm trying to integrate a C425 with Home Assistant but I'm not getting motion detection events after having read through the items in the FAQ. I see you mention webhooks are enabled here. Are you able to get motion detection events from the C425 via webhooks?

**Comment by marcosngomezi** [comment]
the connection Information gives error on hub cameras, maybe useful to avoid the child check or to complement it

**Comment by murtaugh82** [comment]
> In any case, integration does load up to date data every 30s in one request, we can tone it down, ideally automatically. 

Hi @JurajNyiri , i would like to change the update interval manually for the time being until there is another solution. However, i could not identify the corresponding place in the code. Could you please give me a hint where the update is triggered?

**Comment by JurajNyiri** [comment]
Hi, I do not own battery operated camera so it is very hard for me to develop this change without having access to one.

That being said, I tried something.

Could you please download https://github.com/JurajNyiri/HomeAssistant-Tapo-Control/tree/experimental_battery_optimization manually and see what warnings do you get in the logs? This is a POC version, it changes update interval for powered devices to 1 minute and battery powered to 10, but as I said, I had no way to test, so it might not work... 
<img width="914" alt="Screenshot 2024-07-04 at 23 52 50" src="https://github.com/JurajNyiri/HomeAssistant-Tapo-Control/assets/5785999/85ccd285-c8a7-4162-adba-3487d1e191a4">

This is what I see for my mains powered devices, let me know if you see something similar (but with longer update interval) for battery cameras.

**Comment by slartibartfast11** [comment]
Thanks, but I think a custom poll time might be required so users of the integration can adjust it to poll the camera _just_ often enough to conserve the battery but also to not let it fully go to sleep. 

Looking at my C425 camera, it seems like it uses a proprietary wake-up protocol with TP-Link and the official app, which gives the camera just enough connectivity to wait for the wake-up event from TP-Link cloud. Once the camera has gone asleep, there's nothing other than the app (or, obviously, a PIR motion event) that will wake it.

This assumes TP-Link are not willing to come to the party to release a public API to their app wireless camera wake up protocol, of course...

**Comment by JurajNyiri** [comment]
Yes that is the plan. I need the poc tested first though to ensure the logic to differentiate cameras work.

**Comment by murtaugh82** [comment]
Hi @JurajNyiri,

many thanks for the adjustment! I have a C410 and operate it with a solar panel. Under _basic_info_ there is an attribute _power_mode_ (with value _SOLAR_ or _BATTERY_) instead of _power_ for this model.

```
2024-07-05 12:23:52.639 DEBUG (MainThread) [custom_components.tapo_control] {'getDeviceInfo': [{'device_info': {'basic_info': {'device_type': 'SMART.IPCAMERA', 'device_info': 'C410 1.0 IPC', 'features': 3, 'barcode': '', 'device_model': 'C410', 'sw_version': '1.0.16 Build 240430 Rel.63325n', 'device_name': 'C410 1.0', 'hw_version': '1.0', 'device_alias': 'Tapo_C410_Backyard', 'mobile_access': '0', 'mac': 'xxx', 'dev_id': 'xxx', 'hw_id': 'xxx', 'oem_id': 'xxx', 'hw_desc': '00000000000000000000000000000000', 'power_mode': 'SOLAR', 'battery_percent': 100, 'battery_charging': 'NO', 'battery_overheated': False, 'power_save_mode': 'off', 'power_save_status': 'off', 'low_battery': False, 'manufacturer_name': 'TP-LINK', 'ffs': False, 'is_cal': True, 'avatar': '', 'has_set_location_info': 1, 'longitude': xxx, 'latitude': xxx}}}] ...
```
I have therefore adapted the code a little (sorry for the mess - I have neither pyhton experience nor a proper ide :wink: )...

```
controllerData["isRunningOnBattery"] = (
	True
	if (
		"basic_info"
		in updateDataForAllControllers[controller]
		and (
				(
					"power"
					in updateDataForAllControllers[controller][
						"basic_info"
					]
					and updateDataForAllControllers[controller][
						"basic_info"
					]["power"]
					== "BATTERY"
				)  or  (
					"power_mode"
					in updateDataForAllControllers[controller][
						"basic_info"
					]
					and (
						(
							updateDataForAllControllers[controller][
								"basic_info"
							]["power_mode"]
							== "BATTERY"                                        
						) or (
							updateDataForAllControllers[controller][
								"basic_info"
							]["power_mode"]
							== "SOLAR"                                                      
						)
					)
				)
		)
	)
	else False
```

...and now get the desired behavior :smile:

```
2024-07-05 13:04:50.811 WARNING (MainThread) [custom_components.tapo_control] Tapo_C410_Terrace running on battery: False
2024-07-05 13:04:50.811 WARNING (MainThread) [custom_components.tapo_control] Updating Tapo_C410_Terrace...
2024-07-05 13:04:52.072 WARNING (MainThread) [custom_components.tapo_control] Tapo_C410_Backyard running on battery: False
2024-07-05 13:04:52.072 WARNING (MainThread) [custom_components.tapo_control] Updating Tapo_C410_Backyard...
2024-07-05 13:04:54.157 WARNING (MainThread) [homeassistant.components.select] Setup of select platform tapo_control is taking over 10 seconds.
2024-07-05 13:05:02.067 WARNING (MainThread) [custom_components.tapo_control] Tapo_C410_Terrace running on battery: True
2024-07-05 13:05:02.067 WARNING (MainThread) [custom_components.tapo_control] Skipping update for Tapo_C410_Terrace...
2024-07-05 13:05:03.053 WARNING (MainThread) [custom_components.tapo_control] Tapo_C410_Backyard running on battery: True
2024-07-05 13:05:03.053 WARNING (MainThread) [custom_components.tapo_control] Skipping update for Tapo_C410_Backyard...
2024-07-05 13:05:04.205 WARNING (MainThread) [homeassistant.helpers.frame] Detected that custom integration 'hacs' accesses hass.helpers.event. This is deprecated and will stop working in Home Assistant 2024.11, it should be updated to import functions used from event directly at custom_components/hacs/base.py, line 642: self.hass.helpers.event.async_track_time_interval(, please create a bug report at https://github.com/hacs/integration/issues
2024-07-05 13:05:14.171 WARNING (MainThread) [custom_components.tapo_control] Tapo_C410_Entrance running on battery: False
2024-07-05 13:05:14.171 WARNING (MainThread) [custom_components.tapo_control] Updating Tapo_C410_Entrance...
2024-07-05 13:05:20.791 WARNING (MainThread) [custom_components.tapo_control] Tapo_C410_Terrace running on battery: True
2024-07-05 13:05:20.791 WARNING (MainThread) [custom_components.tapo_control] Skipping update for Tapo_C410_Terrace...
2024-07-05 13:05:21.768 WARNING (MainThread) [custom_components.tapo_control] Tapo_C410_Backyard running on battery: True
2024-07-05 13:05:21.768 WARNING (MainThread) [custom_components.tapo_control] Skipping update for Tapo_C410_Backyard...
2024-07-05 13:05:24.171 WARNING (MainThread) [homeassistant.helpers.entity] Update of sensor.tapo_c410_entrance_battery is taking over 10 seconds
2024-07-05 13:05:27.844 WARNING (MainThread) [homeassistant.helpers.entity] Update of camera.tapo_c410_entrance_hd_stream is taking over 10 seconds
2024-07-05 13:05:31.079 WARNING (MainThread) [custom_components.tapo_control] Tapo_C410_Terrace running on battery: True
2024-07-05 13:05:31.079 WARNING (MainThread) [custom_components.tapo_control] Skipping update for Tapo_C410_Terrace...
2024-07-05 13:05:32.205 WARNING (MainThread) [custom_components.tapo_control] Tapo_C410_Backyard running on battery: True
2024-07-05 13:05:32.205 WARNING (MainThread) [custom_components.tapo_control] Skipping update for Tapo_C410_Backyard...
2024-07-05 13:05:37.856 WARNING (MainThread) [custom_components.tapo_control] Tapo_C410_Entrance running on battery: True
2024-07-05 13:05:37.856 WARNING (MainThread) [custom_components.tapo_control] Skipping update for Tapo_C410_Entrance...

....

2024-07-05 13:15:00.828 WARNING (MainThread) [custom_components.tapo_control] Updating Tapo_C410_Terrace...
2024-07-05 13:15:02.068 WARNING (MainThread) [custom_components.tapo_control] Tapo_C410_Backyard running on battery: True
2024-07-05 13:15:02.069 WARNING (MainThread) [custom_components.tapo_control] Updating Tapo_C410_Backyard...
2024-07-05 13:15:07.956 WARNING (MainThread) [custom_components.tapo_control] Tapo_C410_Entrance running on battery: True
2024-07-05 13:15:07.957 WARNING (MainThread) [custom_components.tapo_control] Skipping update for Tapo_C410_Entrance...
2024-07-05 13:15:12.085 WARNING (MainThread) [custom_components.tapo_control] Tapo_C410_Terrace running on battery: True
2024-07-05 13:15:12.086 WARNING (MainThread) [custom_components.tapo_control] Skipping update for Tapo_C410_Terrace...
2024-07-05 13:15:12.994 WARNING (MainThread) [custom_components.tapo_control] Tapo_C410_Backyard running on battery: True
2024-07-05 13:15:12.994 WARNING (MainThread) [custom_components.tapo_control] Skipping update for Tapo_C410_Backyard...
2024-07-05 13:15:17.959 WARNING (MainThread) [custom_components.tapo_control] Tapo_C410_Entrance running on battery: True
2024-07-05 13:15:17.959 WARNING (MainThread) [custom_components.tapo_control] Updating Tapo_C410_Entrance...

```

**Comment by JurajNyiri** [comment]
@murtaugh82 that is great!

- Could you please confirm it actually improves the battery life vs the currently released version?
- Could you test a longer time to see if the camera wakes up correctly (see what the max value is, to see how high we can go) - just change https://github.com/JurajNyiri/HomeAssistant-Tapo-Control/blob/experimental_battery_optimization/custom_components/tapo_control/const.py#L58

You can do second point along with the first one, so that we know what the recommended value should be for a good balance of battery life vs updates in HA.

**Comment by murtaugh82** [comment]
I have set the update interval to 30 minutes and the logs also show that the update is only performed every 30 minutes. Unfortunately, this does not reduce the discharge of the battery.

![grafik](https://github.com/JurajNyiri/HomeAssistant-Tapo-Control/assets/170609601/5bc19324-0dda-4c0c-a852-d230dafca811)

Is it possible that the connection to the camera remains active even there is no update? with the old version of the integration, i could see that the status led was always on until i deactivated the integration. 

Unfortunately, I can't check this at the moment as my working cameras are over 1000km away. I have two more devices on site, but they are no longer accessible due to a firmware update (Tapo C410(EU)_V1_1.0.17 Build 240524). So that I can test them, these cameras need to be downgraded to the **Tapo C410(EU)_V1_1.0.16 Build 240430** firmware. However, **I don't currently know how to get this specific version and would be grateful for any help.**

**Comment by JurajNyiri** [comment]
Here https://github.com/JurajNyiri/HomeAssistant-Tapo-Control/issues/457#issuecomment-2211332674  https://github.com/JurajNyiri/HomeAssistant-Tapo-Control/blob/experimental_battery_optimization/custom_components/tapo_control/const.py#L58

**Comment by murtaugh82** [comment]
I had already set this value to 30 * 60 and the update messages were then only logged every 30 minutes. (see https://github.com/JurajNyiri/HomeAssistant-Tapo-Control/issues/457#issuecomment-2210746714)

However, I have now noticed that the pytapo request logs are also appearing but at much shorter intervals.

**Comment by murtaugh82** [comment]
The next attempt looks more promising. Within about 24 hours, the cameras have lost only 1%.

Update interval: 30 minutes
Camera streams: deactivated

![grafik](https://github.com/user-attachments/assets/3472afe9-e767-4d99-a269-531bc3a04eb3)

**Comment by JurajNyiri** [comment]
Implemented in https://github.com/JurajNyiri/HomeAssistant-Tapo-Control/releases/tag/5.4.27 . Thank you @murtaugh82 !

**Comment by JurajNyiri** [comment]
Going to reopen to add the solar part

**Comment by JurajNyiri** [comment]
Released in https://github.com/JurajNyiri/HomeAssistant-Tapo-Control/releases/tag/5.4.28.

@murtaugh82 please let me know if it works as expected

**Comment by JurajNyiri** [comment]
Reopening again to make this user configurable.

**Comment by JurajNyiri** [comment]
Update interval is now configurable in https://github.com/JurajNyiri/HomeAssistant-Tapo-Control/releases/tag/5.4.29 .

<img width="433" alt="Screenshot 2024-07-20 at 13 38 44" src="https://github.com/user-attachments/assets/6072750a-4ae4-4160-afe5-e36ec68ca960">

**Comment by murtaugh82** [comment]
> @murtaugh82 please let me know if it works as expected

feature works as expected. thx 😃

**Comment by carontony** [comment]
> Hi @murtaugh82,
> 
> I have the same C410 model which then having the same issue that latest firmware 1.0.17 caused it can't access locally. Not sure how you managed to fix it ?

You can maybe follow this informations to downgrade your firmware : https://sizeof.cat/post/tp-link-tapo-c310-camera/

Let us know.

**Comment by corazzam** [comment]
Hi, I have the C425. I have the same issue of the battery draining too quickly.. I tried changing the refresh rate to an higher delay but it didn’t work. Anyone experiencing the same issue with the C425 (latest firmware 1.2.25)?

**Comment by takalti-ha** [comment]
C410 same issue. would love a fix. As a side note, my ring cameras do not do this.

**Comment by OPUM-LABS** [comment]
I know this issue has been closed, but I wanted to follow up on it anyway because I think I've found the solution, as I had the same problem (overnight consumption now: up to 10%):

In my case, the activated “Media Sync” option was responsible.

I have now solved this with an automation, whereby Media Sync is activated for 20 minutes at 8 a.m. and then deactivated again. (Overnight consumption now: 1-2%)

It also seems that the camera charges faster (less consumption during charging).

Automation:
<img width="1079" height="746" alt="Image" src="https://github.com/user-attachments/assets/deb02f96-0a28-4135-9296-912bda742a48" />

**Comment by natterangell** [comment]
Just chiming in that I experienced the same issue as @OPUM-LABS with regards to battery drain. Went from 54% to 17% over night. About 3 and a half hours of usage on my C420 in the Tapo app. Disabled media sync and the problem is solved. I initially thought I had made some mistake setting up the solar panel, but after adjusting this it charges just fine. I've now created a similar automation to trigger media sync for 20 minutes in the morning to fetch any recordings. This is fine for me with this camera.
